### PR TITLE
Add support for SCRAM when using SASL authN

### PR DESCRIPTION
**Pull request by eloparco**
When using SASL authentication, currently only Kerberos mechanism is supported.
It should be possible to also use SCRAM (e.g. `"SCRAM-SHA-256"` as `"sasl.mechanism"`) without incurring an exception.

In general, SASL should support both Kerberos and SCRAM as strong authentication mechanisms.

### CODE REVIEW

File: include/kafka/KafkaClient.h [hunk 1]
```diff
@@ -376,15 +377,17 @@ KafkaClient::validateAndReformProperties(const Properties& origProperties)
...
         {
-            if (!properties.getProperty(SASL_KERBEROS_SERVICE_NAME))
+            auto saslMechanism = properties.getProperty(SASL_MECHANISM);
+            if (!properties.getProperty(SASL_KERBEROS_SERVICE_NAME) &&
+                saslMechanism->find("SCRAM") == std::string::npos)
             {
                 KAFKA_THROW_WITH_MSG(RD_KAFKA_RESP_ERR__INVALID_ARG,\
-                                     "The \"sasl.kerberos.service.name\" property is mandatory for SASL connection!");
+                                     "Either the \"sasl.kerberos.service.name\" property or the \"sasl.mechanism\" set to SCRAM must be provided for SASL connection!");
             }
```
Comment (kenneth-jia) [inline include/kafka/KafkaClient.h:391]: Thank you for pointing the issue out. 
Then I guess we don't need the config check here, -- just remove all these lines would be OK. 😊

File: include/kafka/KafkaClient.h [hunk 2]
```diff
@@ -198,6 +198,7 @@ class KafkaClient
     static const constexpr char* DEBUG             = "debug";
     static const constexpr char* SECURITY_PROTOCOL          = "security.protocol";
     static const constexpr char* SASL_KERBEROS_SERVICE_NAME = "sasl.kerberos.service.name";
```
Comment (kenneth-jia) [inline include/kafka/KafkaClient.h:200]: Then we don't need these 2 lines either.

**Comment by eloparco** [comment]
It could still be useful to have some constraints to avoid incomplete configurations (SASL without specifying Kerberos or SCRAM) or weak configurations (SASL + PLAIN).
What do you think?

**Comment by kenneth-jia** [comment]
> It could still be useful to have some constraints to avoid incomplete configurations (SASL without specifying Kerberos or SCRAM) or weak configurations (SASL + PLAIN).
> What do you think?
Actually I'm thinking about providing some error notification interface, or throw an exception for rejected properties, -- there's already config-verification in `librdkafka`, thus we could keep it simple in `modern-cpp-kafka`.

**Comment by kenneth-jia** [comment]
@eloparco I raised a PR https://github.com/morganstanley/modern-cpp-kafka/pull/86 to remove the SASL configuration checking, since the check would not help too much.
A better way would be registering an error callback (i.e. `setErrorCallback`), thus we would be able to get detailed error info.
E.g,
```
Local: Authentication failure [-169] | sasl_plaintext://.../bootstrap: Failed to initialize SASL authentication: SASL handshake failed (start (-1)): SASL(-1): generic failure: GSSAPI Error: Unspecified GSS failure.  Minor code may provide more information (Server .... not found in Kerberos database) ...
```

**Comment by kenneth-jia** [comment]
Close the PR since #86 has been merged.

**Pull request closed without merging**